### DISCUSSION TITLE
Create quarto reports and branching - different useful names of reports?

### DISCUSSION THREAD

**Discussion by rkrug** [ropensci/targets#1297] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

This is a follow up to https://github.com/ropensci/targets/discussions/1296 - in a nutshell, I do dynamic branding over files and create for each file graphs and some summary statistics.

Now I would like to have at the end have a report which is based on measures calculated as a target as well as the graphs created. I have e.g. `tar_read(figure_oa_status_ias)` in the quarto document to include a graph, `tar_read(bibliography)` to read the data. How can I do this when I am using dynamic branching?

my pipeline looks as follows:

```
tar_files(
        files,
        c(
            file.path(".", "input", "IPBES ILK.csv"),
            file.path(".", "input", "IPBES IAS.csv")
        )
    ),
    tar_target(bibliography, load_bibliography(files), pattern = map(files)),
    tar_target(figure_pub_year, plot_publication_year(bibliography), pattern = map(bibliography)),
    tar_target(figure_oa_status, plot_oa_status(bibliography), pattern = map(bibliography)),
    tar_target(figure_top_journ_data, plot_top_journals_data(bibliography), pattern = map(bibliography)),
    tar_target(figure_top_journals, plot_top_journals(figure_top_journ_data), pattern = map(figure_top_journ_data)),
    tar_target(figure_top_country_data, plot_top_country_data(bibliography), pattern = map(bibliography)),
    tar_target(figure_top_country, plot_top_country(figure_top_country_data), pattern = map(figure_top_country_data))
```

and I would like to add something along the lines of

```
tar_quarto(
        report,
        "bibliography_report.qmd",
        quiet = FALSE,
        map = (????)
    )
```

So that at the end I have one report for each branch.

**Comment by rkrug** [comment]
I switched to Static Branching because of the file names - makes more sense in this context.

I now have:
```
values <- data.frame(
    files = list.files(
        path = "input",
        pattern = "\\.csv$",
        full.names = FALSE
    ) |>
        gsub(pattern = "\\.csv$", replacement = "")
    # c(
    #     ILK = file.path(".", "input", "IPBES ILK.csv"),
    #     IAS = file.path(".", "input", "IPBES IAS.csv")
    # )
)
tar_map(
    values = values,
    tar_target(bibliography, load_bibliography(files)),
    # tar_target(metrics, dois_measures(bibliography)),
    tar_target(figure_pub_year, plot_publication_year(bibliography)),
    tar_target(figure_oa_status, plot_oa_status(bibliography)),
    tar_target(figure_top_journ_data, plot_top_journals_data(bibliography)),
    tar_target(figure_top_journals, plot_top_journals(figure_top_journ_data)),
    tar_target(figure_top_country_data, plot_top_country_data(bibliography)),
    tar_target(figure_top_country, plot_top_country(figure_top_country_data)),
    tar_quarto(
        report,
        "bibliography_report.qmd",
        quiet = FALSE
    )
)
```

And everything except of the `tar_quarto()` works. It aborts when trying the first `tar_read()` command, i.e. `bibliography <- targets::tar_read(bibliography)` which makes quite sense, as it is trying to read `bibliography` which does not exist.

**Reply by ginolhac** [reply]
You should have `tar_quarto()` outside of `tar_map()` since it is only one file. Something like

``` r
mapped <- tar_map(
  values = values,
  tar_target(bibliography, load_bibliography(files)),
  # tar_target(metrics, dois_measures(bibliography)),
  tar_target(figure_pub_year, plot_publication_year(bibliography)),
  tar_target(figure_oa_status, plot_oa_status(bibliography)),
  tar_target(figure_top_journ_data, plot_top_journals_data(bibliography)),
  tar_target(figure_top_journals, plot_top_journals(figure_top_journ_data)),
  tar_target(figure_top_country_data, plot_top_country_data(bibliography)),
  tar_target(figure_top_country, plot_top_country(figure_top_country_data)),
  
)

list(mappped, tar_quarto(
  report,
  "bibliography_report.qmd",
  quiet = FALSE
))
```

`{targets}` wants a list eventually, you can define it with intermediate ones if you wish

**Reply by rkrug** [reply]
Sorry for not being clear - I would like to have one report per branch. The summary report will come after that.

**Reply by ginolhac** [reply]
then, you can look at this example here: https://github.com/ropensci/tarchetypes/discussions/175#discussioncomment-9675612, because there are [literate programming limitations](https://docs.ropensci.org/tarchetypes/reference/tar_quarto.html#literate-programming-limitations)

**Reply by rkrug** [reply]
Thanks @ginolhac - I expected something like that. I will try the following:

1. give the targets as parameter to quarto, which should change them in the `tar_map()` function()
My only question. is then how can I set the output file name to the name of the branch?

**Reply by ginolhac** [reply]
in the previous linked example I have something like, where now you can add the `description` field so it shows up in both graph and manifest.

``` r
library(targets)
library(tarchetypes)

values <- list(name = rlang::syms(c("report1", "report2", "report3")), 
               file = c("report1.Rmd", "report2.Rmd", "report3.Rmd"))

list(
  tar_target(params, data.frame(par = c("one", "two", "three"),
                                value = c(1, 2, 3))),
  #tar_target(testing, params[sample(nrow(params), 1), ]),
  tar_eval(tar_render(name, file, description = file), values = values)
)
```

![image](https://github.com/ropensci/targets/assets/710181/b1edaea9-27d8-49f4-8038-29c511ddbff5)

Of note, `description` needs targets >= 1.6.0.

**Reply by rkrug** [reply]
But this would require different report templates - correct? and I have one cmd report, which I would like to render for each branch. Could II use instead of `file` the report name? But I assume quarto would overwrite it each time?

**Reply by ginolhac** [reply]
if you use static branching, you should know upfront the different things you iterate over. So you should have different reports.
Or you can use parametrized reports if the skeleton is almost identical

**Reply by rkrug** [reply]
I think parametrised report is the way to go. The easiest would be to give the branch name (the part after the `_` and then in the report use something along the lines of `tar_read(bibliography_THE_BRANCH_NAME)` but I tried this already - and did not succeed. How can I make `tar_read(paste0("bibliography_", "IPBES.IAS"))` read the target named `bibliography_IPBES.IAS`? I think then I have it (mainly).

**Reply by ginolhac** [reply]
you can use `tar_name()` that returns the name of the current branch as character. You should maybe provide a reprex so we can test it for real

**Reply by rkrug** [reply]
Yes - but I am getting somewhere. I now have the branch name in the quarto file as a parameter 

```
values <- data.frame(
    files = list.files(
        path = "input",
        pattern = "\\.csv$",
        full.names = FALSE
    ) |>
        gsub(pattern = "\\.csv$", replacement = "")
)
names <- data.frame(
    name = gsub(values$files, pattern = " ", replacement = ".")
)
list(
    tar_map(
        values = values,
        tar_target(bibliography, load_bibliography(files)),
        tar_target(metrics, get_bibliography_measures(bibliography)),
        tar_target(figure_pub_year, plot_publication_year(bibliography)),
        tar_target(figure_oa_status, plot_oa_status(bibliography)),
        tar_target(figure_top_journ_data, plot_top_journals_data(bibliography)),
        tar_target(figure_top_journals, plot_top_journals(figure_top_journ_data)),
        tar_target(figure_top_country_data, plot_top_country_data(bibliography)),
        tar_target(figure_top_country, plot_top_country(figure_top_country_data))
    ),
    tar_eval(
        tar_quarto(
            name = name,
            path = "bibliography_report.qmd",
            execute_params = list(branch = tar_name())
        ),
        values = names
    )
)
```

and I read the targets using `tar_read_raw()` which works b ut gives a waring (not nice, but I can live with this).

My last 2 problems: 

1. The `tar_quarto()` target  is executed directly after the bibliography - which does not work as it depends on the other targets. Can I force it to run last?
2. The second time it runs, it overwrites the first one, and I do not manage to specify the output file name of the quarto rendering. I can not specify `output` in `tar_quarto()` and when I use `quarto_args = 'output_file = your_output_filename.html' ` I get a panic error. I lean towards making my own target which calls `quarto_render()` and at the end remains the resulting report to what I want it to be.

**Reply by ginolhac** [reply]
Hi,
I had to work hard on this one, simplify it to something easier.
You really want static branching while the dynamic one would be clearer in the dependency graph. Yes branch names are more complex but now we need a `tar_combine()` step. Nevertheless, see below the sketch for static + parametrized quarto docs.
And I think the best for **parametrized** quarto docs is to use [`tar_quarto_rep()`](https://docs.ropensci.org/tarchetypes/reference/tar_quarto_rep.html).

`report.qmd`:

~~~ r
---
title: "biblio"
format: html
params:
  name: "default_name"
  value: "default_val"
---


## discussion 1297
```

```{r}
print(params$name)
```

~~~

`_targets.R` file

``` r
library(targets)
library(tarchetypes)

values <- data.frame(
  val = 1:3,
  names = c("A", "B", "C")
)

mapped <- tar_map(
  values = values,
  tar_target(bibliography, paste(val, "bib")),
  tar_target(summary, paste(bibliography, "sum"))
)

list(
  mapped,
  tar_combine(combi, mapped[["summary"]], 
              command = dplyr::bind_rows(!!!.x) |> 
                tidyr::pivot_longer(cols = dplyr::everything()) |> 
                dplyr::mutate(output_file = paste0(name, ".html"))),
  tar_quarto_rep(
    name = report,
    path = "report.qmd",
    execute_params = combi
  )
)
```

DAG:
![image](https://github.com/ropensci/targets/assets/710181/afe2885e-3a09-4943-8726-2dee2d0493b5)


After `tar_make()` the folder contains the 3 rendered HTML files using the provided `output_file` in `tar_combine()`

```
report_files
report.qmd
summary_1_A.html
summary_2_B.html
summary_3_C.html
_targets
_targets.R
```

Example of `summary_1_A.html`
![image](https://github.com/ropensci/targets/assets/710181/3db0c491-8cfb-4d45-b49a-a576be9f5a98)

`summary_2_B.html`
![image](https://github.com/ropensci/targets/assets/710181/67db08af-5641-4a6b-af37-e1fead3bfb11)

**Comment by rkrug** [comment]
Thanks - and sorry for causing that much work for you. I think I am out of my depth now (dplyr was never my strong side). 

Final question: You said that dynamic branching would be better in this case - how would your example look in the case of dynamic branching? Maybe I get used to the hashed branch names...

**Reply by ginolhac** [reply]
No problem, I like playing with this package.
To ease yourself and also others, doing **reprex** is great. By simplifying, things usually become clearer and others can run code on their machine. 
See this reprex for your last demand, reading biblio csv (fake but created inside the reprex) in a **dynamic** context.

``` r
# Parameterized Quarto:
lines <- c(
  "---",
  "title: 'biblio'",
  "format: html",
  "params:",
  "  author: \"default value\"",
  "  value: \"default value\"",
  "---",
  "## discussion 1297",
  "```{r}",
  "print(params$author)",
  "print(params$value)",
  "```"
)
writeLines(lines, con = "report.qmd")
# fake files
writeLines(c("author,year", "a,2012"), con = "bib1.csv")
writeLines(c("author,year", "bb,2014", "bb,2015"), con = "bib2.csv")
writeLines(c("author,year", "ac,2016", "ac,2010", "ac,1998"), con = "bib3.csv")
# the dynamic pipeline
targets::tar_script({
  library(tarchetypes)
  list(
    tar_files_input(bibfiles, fs::dir_ls(glob = "*.csv")),
    tar_target(bib, readr::read_csv(bibfiles, show_col_types = FALSE) |> 
                 # create the variable for the output filename based on the 
                 # input files,
                 dplyr::mutate(output_file = stringr::str_replace(bibfiles, ".csv$", ".html")),
               pattern = map(bibfiles)),
    tar_target(bibcount, dplyr::count(bib, output_file, author, name = "value"),
               pattern = map(bib)),
    tar_quarto_rep(
      report,
      "report.qmd",
      execute_params = bibcount
    )
  )
}, ask = FALSE)
```
```r
> targets::tar_make()
▶ dispatched target bibfiles_files
● completed target bibfiles_files [0.001 seconds]
▶ dispatched branch bibfiles_a854fcb2edcbcc69
● completed branch bibfiles_a854fcb2edcbcc69 [0 seconds]
▶ dispatched branch bibfiles_37dc96391e254039
● completed branch bibfiles_37dc96391e254039 [0 seconds]
▶ dispatched branch bibfiles_dd5ecd4757c5539f
● completed branch bibfiles_dd5ecd4757c5539f [0 seconds]
● completed pattern bibfiles
▶ dispatched branch bib_009cfea8ffb7cde0
● completed branch bib_009cfea8ffb7cde0 [0.253 seconds]
▶ dispatched branch bib_463491e38f68f4fd
● completed branch bib_463491e38f68f4fd [0.006 seconds]
▶ dispatched branch bib_7ab6f4a25026850d
● completed branch bib_7ab6f4a25026850d [0.006 seconds]
● completed pattern bib
▶ dispatched branch bibcount_bac1924fb7edecfc
● completed branch bibcount_bac1924fb7edecfc [0.026 seconds]
▶ dispatched branch bibcount_60d955e73ee0b234
● completed branch bibcount_60d955e73ee0b234 [0.006 seconds]
▶ dispatched branch bibcount_62ba7581982a4a01
● completed branch bibcount_62ba7581982a4a01 [0.008 seconds]
● completed pattern bibcount
▶ dispatched target report_params
● completed target report_params [0.002 seconds]
▶ dispatched branch report_b8aa2b8c7e9d8782
● completed branch report_b8aa2b8c7e9d8782 [2.564 seconds]
▶ dispatched branch report_b769eb575a9413a0
● completed branch report_b769eb575a9413a0 [2.743 seconds]
▶ dispatched branch report_f1fab0c1bf797437
● completed branch report_f1fab0c1bf797437 [2.677 seconds]
● completed pattern report
▶ ended pipeline [8.69 seconds]
```
``` r
targets::tar_visnetwork(label = "branches")
```
![image](https://github.com/ropensci/targets/assets/710181/7f1d1978-9efc-4b92-b658-83bb1720d9cd)

the files once done:

```
❯ ls -1
bib1.csv
bib1.html
bib2.csv
bib2.html
bib3.csv
bib3.html
dynamic.R
report_files
report.qmd
_targets
_targets.R
```

- first report:
![image](https://github.com/ropensci/targets/assets/710181/07161c2f-1cd9-4b84-b95e-6d218eee1b84)

- second report
![image](https://github.com/ropensci/targets/assets/710181/344f3f1a-f7bb-429a-a0eb-9feb9b0418b5)

Branch names are cryptic but the filenames are clear and the DAG looks just great as one line
